packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2957,29 +2972,76 @@ export class ContainerRuntime
 			) {
-				// only logging for the first connection and the trackingSignalSequenceNUmber.
-				if (this.consecutiveReconnects === 0) {
-					this.sendSignalTelemetryEvent(envelope.clientSignalSequenceNumber);
+				if (
+					envelope.clientSignalSequenceNumber >
+					this._signalTracking.trackingSignalSequenceNumber
+				) {
+					// Incoming signal sequence number is greater than the expected signal sequence number.
+					// So we calculate the number of signals lost and log the event.
+					const signalsLost =
+						envelope.clientSignalSequenceNumber -
+						this._signalTracking.trackingSignalSequenceNumber;
+					this._signalTracking.signalsLost += signalsLost;
+					this.mc.logger.sendErrorEvent({
+						eventName: "SignalLost",
+						type: envelope.contents.type,
+						signalsLost,
+						trackingSequenceNumber: this._signalTracking.trackingSignalSequenceNumber,
+						clientSignalSequenceNumber: envelope.clientSignalSequenceNumber,
+					});
+					// Update the tracking signal sequence number to the next expected signal in the sequence.
+					this._signalTracking.trackingSignalSequenceNumber =
+						envelope.clientSignalSequenceNumber + 1;
+				} else if (
+					envelope.clientSignalSequenceNumber <
+						this._signalTracking.trackingSignalSequenceNumber &&
+					envelope.clientSignalSequenceNumber >=
+						this._signalTracking.minimumTrackingSignalSequenceNumber
+				) {
+					if (
+						this._signalTracking.previousRoundTripSignalSequenceNumber !== undefined &&
+						envelope.clientSignalSequenceNumber >
+							this._signalTracking.previousRoundTripSignalSequenceNumber
+					) {
+						// Log incoming signal that was previously marked as lost as having been received out of order.
+						this._signalTracking.signalsLost--;
+					}
+					this.mc.logger.sendErrorEvent({
+						eventName: "SignalOutOfOrder",
+						type: envelope.contents.type,
+						trackingSequenceNumber: this._signalTracking.trackingSignalSequenceNumber,
+						clientSignalSequenceNumber: envelope.clientSignalSequenceNumber,
+					});
+				} else if (
+					envelope.clientSignalSequenceNumber ===
+					this._signalTracking.trackingSignalSequenceNumber
+				) {
+					// Incoming signal is the next expected signal in the sequence.
+					// Update the tracking signal sequence number to the next expected signal.
+					this._signalTracking.trackingSignalSequenceNumber++;
+				}
+
+				if (
+					this._signalTracking.roundTripSignalSequenceNumber !== undefined &&
+					this._signalTracking.roundTripSignalSequenceNumber <=
+						envelope.clientSignalSequenceNumber
+				) {
+					if (
+						this.consecutiveReconnects === 0 &&
+						envelope.clientSignalSequenceNumber ===
+							this._signalTracking.roundTripSignalSequenceNumber
+					) {
+						this.sendSignalTelemetryEvent(envelope.clientSignalSequenceNumber);
+					} else {
+						this._signalTracking.trackingSignalSequenceNumber =
+							envelope.clientSignalSequenceNumber + 1;
```

Comment:
Seem to be setting this in two different places.

Reply:
Yeah, this is redundant. This line has been removed.